base/utf16.jl

```diff
@@ -20,92 +45,248 @@ function endof(s::UTF16String)
+            # output surrogate pair
+            buf[out += 1] = UInt16(0xd7c0 + (ch >>> 10))
+            buf[out += 1] = UInt16(0xdc00 + (ch & 0x3ff))
         end
     end
-    push!(buf, 0) # NULL termination
+    @inbounds buf[out + 1] = 0 # NULL termination
     UTF16String(buf)
 end
 
-utf16(x) = convert(UTF16String, x)
-convert(::Type{UTF16String}, s::UTF16String) = s
-convert(::Type{UTF16String}, s::AbstractString) = encode16(s)
-convert(::Type{Array{UInt16,1}}, s::UTF16String) = s.data
-convert(::Type{Array{UInt16}}, s::UTF16String) = s.data
+"
+Converts a UTF8String to a UTF16String
 
-# TODO: optimize this
-convert(::Type{UTF8String}, s::UTF16String) =
-    sprint(length(s.data)-1, io->for c in s; write(io,c::Char); end)
+### Input Arguments:
+*   ::Type{UTF16String}
+*   str::UTF8String
 
-sizeof(s::UTF16String) = sizeof(s.data) - sizeof(UInt16)
-unsafe_convert{T<:Union(Int16,UInt16)}(::Type{Ptr{T}}, s::UTF16String) =
-    convert(Ptr{T}, pointer(s))
+### Returns:
+*   UTF16String
 
-function isvalid(::Type{UTF16String}, data::AbstractArray{UInt16})
-    i = 1
-    n = length(data) # this may include NULL termination; that's okay
-    while i < n # check for unpaired surrogates
-        if utf16_is_lead(data[i]) && utf16_is_trail(data[i+1])
-            i += 2
-        elseif utf16_is_surrogate(data[i])
-            return false
+### Throws:
+*   UnicodeError
+"
+function convert(::Type{UTF16String}, str::UTF8String)
+    dat = str.data
+    # handle zero length string quickly
+    sizeof(dat) == 0 && return empty_utf16
+    # Check that is correct UTF-8 encoding and get number of words needed
+    len, flags, num4byte = check_string(dat)
+    len += num4byte
+    buf = Vector{UInt16}(len+1)
+    @inbounds buf[len+1] = 0
+    # Optimize case where no characters > 0x7f
+    flags == 0 && @inbounds return UTF16String(copy!(buf, dat))
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch::UInt32 = dat[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle range 0x80-0x7ff
+        elseif ch < 0xe0
+            buf[out += 1] = ((ch & 0x1f) << 6) | (dat[pos += 1] & 0x3f)
+        # Handle range 0x800-0xffff
+        elseif ch < 0xf0
+            pos += 2
+            buf[out += 1] = get_utf8_3byte(dat, pos, ch)
+        # Handle range 0x10000-0x10ffff
         else
-            i += 1
+            pos += 3
+            ch = get_utf8_4byte(dat, pos, ch)
+            # output surrogate pair
+            buf[out += 1] = UInt16(0xd7c0 + (ch >>> 10))
+            buf[out += 1] = UInt16(0xdc00 + (ch & 0x3ff))
         end
     end
-    return i > n || !utf16_is_surrogate(data[i])
+    UTF16String(buf)
 end
 
+"
+Converts a UTF-16 encoded vector of UInt16 to a UTF8String
+
+### Input Arguments:
+*   ::Type{UTF8String}
+*   dat::Vector{UInt16}
+
+### Returns:
+*   UTF8String
+
+### Throws:
+*   UnicodeError
+"
+function convert(::Type{UTF8String}, dat::Vector{UInt16})
```

Comment:
Please take a look at @quinnj 's new `Strings.jl` package, in particular at `Encodings.jl`
(esp. the version in my PR for `Strings.jl`).
That is really how I believe strings should be handled.
The current design of strings in Julia conflates the storage representation (i.e. bytes, 16-bit words, 32-bit words), with both the character set (ASCII vs. Unicode), and how those are logically encoded (i.e. UTF8 vs. UTF16 vs. UTF32).  It also mixes in `\0` termination (and not in a consistent fashion), which ends up causing some of what you've seen as code duplication (necessary to handle those inconsistencies).

This isn't "subverting" the abstraction layer, since this is the module that *implements* that abstraction layer.  If you want real examples of subverting the abstraction layer, I have many,
in `datafmt.jl`, `expr.jl`, `LineEdit.jl`, `pointer.jl`, `printf.jl`, `regex.jl`, `REPL.jl`, and `string.jl`, and those are just in Base...